x-pack/plugins/observability_solution/apm/public/components/app/transaction_details/waterfall_with_summary/waterfall_container/waterfall/context/waterfall_context.tsx

```diff
@@ -0,0 +1,120 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Dictionary, groupBy } from 'lodash';
+import React, { PropsWithChildren, useCallback, useMemo, useState } from 'react';
+import useEffectOnce from 'react-use/lib/useEffectOnce';
+import { CriticalPathSegment } from '../../../../../../../../common/critical_path/types';
+import { getCriticalPath } from '../../../../../../../../common/critical_path/get_critical_path';
+import {
+  buildTraceTree,
+  convertTreeToList,
+  IWaterfall,
+  IWaterfallNode,
+  IWaterfallNodeFlatten,
+  updateTraceTreeNode,
+} from '../waterfall_helpers/waterfall_helpers';
+
+export interface WaterfallContextProps {
+  waterfall: IWaterfall;
+  showCriticalPath: boolean;
+  maxLevelOpen: number;
+  isOpen: boolean;
+}
+
+export const WaterfallContext = React.createContext<
+  {
+    criticalPathSegmentsById: Dictionary<CriticalPathSegment[]>;
+    showCriticalPath: boolean;
+    traceList: IWaterfallNodeFlatten[];
+    getErrorCount: (waterfallItemId: string) => number;
+    updateTreeNode: (newTree: IWaterfallNodeFlatten) => void;
+  } & Pick<WaterfallContextProps, 'showCriticalPath'>
+>({
+  criticalPathSegmentsById: {} as Dictionary<CriticalPathSegment[]>,
+  showCriticalPath: false,
+  traceList: [],
+  getErrorCount: () => 0,
+  updateTreeNode: () => undefined,
+});
+
+export function WaterfallContextProvider({
+  showCriticalPath,
+  waterfall,
+  maxLevelOpen,
+  children,
+  isOpen,
+}: PropsWithChildren<WaterfallContextProps>) {
+  const [tree, setTree] = useState<IWaterfallNode | null>(null);
+  const criticalPath = useMemo(
+    () => (showCriticalPath ? getCriticalPath(waterfall) : undefined),
+    [showCriticalPath, waterfall]
+  );
+
+  const criticalPathSegmentsById = useMemo(
+    () => groupBy(criticalPath?.segments, (segment) => segment.item.id),
+    [criticalPath?.segments]
+  );
+
+  const getErrorCount = useCallback(
+    (waterfallItemId) => waterfall.getErrorCount(waterfallItemId),
+    [waterfall]
+  );
+
+  const traceList = useMemo(() => {
+    return convertTreeToList(tree);
+  }, [tree]);
+
+  const updateTreeNode = useCallback(
+    (updatedNode: IWaterfallNodeFlatten) => {
+      if (!tree) return;
+
+      const newTree = updateTraceTreeNode({
+        root: tree,
+        updatedNode,
+        waterfall,
+        path: {
+          criticalPathSegmentsById,
+          showCriticalPath,
+        },
+      });
+
+      if (newTree) {
+        setTree(newTree);
+      }
+    },
+    [criticalPathSegmentsById, showCriticalPath, tree, waterfall]
+  );
+
+  useEffectOnce(() => {
+    const root = buildTraceTree({
+      waterfall,
+      maxLevelOpen,
+      isOpen,
+      path: {
+        criticalPathSegmentsById,
+        showCriticalPath,
+      },
+    });
+
+    setTree(root);
+  });
```

Comment:
why are we using `useEffectOnce` instead of e.g. `useMemo()` or `useRef()`?

Reply:
tree needs to be a state so that `traceList` can be regenerated whenever a node is updated. it's in this `useEffectOnce because the tree needs to be fully generated only once during the rendering cycle.

Reply:
what about `setState(( ) => buildTraceTrace(...))` ? Only because this delays the initial render with a frame.

Reply:
`criticalPathSegmentsById` needs to be loaded before building the tree. This needs to be in a useEffect.